### PR TITLE
ebi-ait/dcp-ingest-central#19: created endpoint for fetching users according to roles

### PR DESCRIPTION
**Pull request by yusra-haider**
Related PR for front-end: https://github.com/ebi-ait/ingest-ui/pull/90

### CODE REVIEW

**Review by jacobwindsor — APPROVED** [review]

File: src/main/java/org/humancellatlas/ingest/user/UserController.java [hunk 1]
```diff
@@ -106,6 +110,13 @@ private Account getCurrentAccount() {
         return account;
     }
 
+    @GetMapping("/list")
+    ResponseEntity<List<Account>> listUsers(
+            @RequestParam("role") Optional<Role> role) {
+        return role.map(value -> ResponseEntity.ok(accountRepository.findAccountByRolesEquals(value))).orElseGet(() -> ResponseEntity.ok(accountRepository.findAll()));
+    }
+
+
```
Comment (aaclan-ebi) [inline src/main/java/org/humancellatlas/ingest/user/UserController.java:119]: I think the endpoint should be just to get the list of wranglers. And only wranglers should have permission to do it.

File: src/main/java/org/humancellatlas/ingest/security/AccountRepository.java [hunk 2]
```diff
@@ -12,4 +12,5 @@
 
     Account findByProviderReference(String providerReference);
 
+    List<Account> findAccountByRolesEquals(Role role);
```
Comment (amnonkhen) [inline src/main/java/org/humancellatlas/ingest/security/AccountRepository.java:15]: 1. Why `Roles` in plural? 
2. Why is `Equals` necessary?

File: src/main/java/org/humancellatlas/ingest/user/UserController.java [hunk 3]
```diff
@@ -106,6 +110,13 @@ private Account getCurrentAccount() {
         return account;
     }
 
+    @GetMapping("/list")
+    ResponseEntity<List<Account>> listUsers(
+            @RequestParam("role") Optional<Role> role) {
+        return role.map(value -> ResponseEntity.ok(accountRepository.findAccountByRolesEquals(value))).orElseGet(() -> ResponseEntity.ok(accountRepository.findAll()));
```
Comment (amnonkhen) [inline src/main/java/org/humancellatlas/ingest/user/UserController.java:116]: I would put the `.orElseGet` bit in its own line.

File: src/main/java/org/humancellatlas/ingest/user/UserController.java (hunk 1 above)
Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java:119]: I think it's good to have a general endpoint that can also be used to fetch users of other roles, if ever needed in the future.

How do I limit permissions of the api endpoint?

File: src/main/java/org/humancellatlas/ingest/security/AccountRepository.java (hunk 2 above)
Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/security/AccountRepository.java]: 1. Because the property is defined as `roles` in `Account.java`: https://github.com/ebi-ait/ingest-core/blob/c19ccf17f62ef5fd9e7fb9221c136e98c5a7b8b1/src/main/java/org/humancellatlas/ingest/security/Account.java#L60
Using just `Role` gives this error
`Caused by: org.springframework.data.mapping.PropertyReferenceException: No property role found for type Account! Did you mean 'roles'?`
2. It isn't, just removed it

File: src/main/java/org/humancellatlas/ingest/user/UserController.java (hunk 1 above)
Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java]: There is a similar code in the AuthenticationController on checking the role of the user who sent the request. 
```
@GetMapping(path="/account", produces=APPLICATION_JSON_UTF8_VALUE)
    public ResponseEntity<Account> getAccount(Authentication authentication) {
        if (authentication.getAuthorities().contains(Role.GUEST)) {
            return ResponseEntity.notFound().build();
        }
        Account account = (Account) authentication.getPrincipal();
        return ResponseEntity.ok().body(account);
    }

```
I'm still not very comfortable with having a public endpoint returning all users info. However, I think there's a larger work needed to design and implement authorisation properly in the core API 🤔   I also noticed there are 2 classes account & user referring to the same entity. That needs to be fixed. But maybe not in this ticket.

Comment (aaclan-ebi) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java]: Hmm, I'll approve it now, as we don't really store a lot of info in the Account entity. But if in case we have to add more info, e.g. email addresses or other we might want to revisit the access to this. If you have more time for this sprint, could we at least just allow users only with wrangler's role to access this endpoint?

**Review by aaclan-ebi — APPROVED** [review]

File: src/main/java/org/humancellatlas/ingest/user/UserController.java [hunk 4]
```diff
@@ -19,12 +20,13 @@
...
 import org.springframework.hateoas.*;
 import org.springframework.hateoas.mvc.ControllerLinkBuilder;
+import org.springframework.http.ResponseEntity;
 import org.springframework.security.core.Authentication;
 import org.springframework.security.core.context.SecurityContextHolder;
-import org.springframework.web.bind.annotation.CrossOrigin;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.ResponseBody;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (amnonkhen) [inline src/main/java/org/humancellatlas/ingest/user/UserController.java:26]: Why did you change from explicit imports to star?

Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java:26]: ah that was automatic clean up via Intellij. Let me revert the changes

File: src/main/java/org/humancellatlas/ingest/user/UserController.java (hunk 1 above)
Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java]: What about doing this via the `SecurityConfig`?

Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java]: How about making use of the `SecurityConfig` here?

File: src/main/java/org/humancellatlas/ingest/user/UserController.java [hunk 5]
```diff
@@ -106,6 +116,17 @@ private Account getCurrentAccount() {
         return account;
     }
 
+    @GetMapping(path = "/list", produces=APPLICATION_JSON_UTF8_VALUE)
```
Comment (amnonkhen) [inline src/main/java/org/humancellatlas/ingest/user/UserController.java:119]: Have you considered using https://www.baeldung.com/spring-security-method-security#method-security?

**Review by amnonkhen — APPROVED** [review]

File: src/main/java/org/humancellatlas/ingest/user/UserController.java (hunk 5 above)
Comment (yusra-haider) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java:119]: This would be a really nice fix. I'll read about how the decorator works. Thanks for the suggestion!

**Review by aaclan-ebi — APPROVED** [review]
Thanks for applying my comment!
